lib/ansible/parsing/yaml/objects.py

```diff
@@ -113,35 +116,264 @@ def __init__(self, ciphertext):
     @property
     def data(self):
         if not self.vault:
-            return self._ciphertext
+            return to_text(self._ciphertext)
         return to_text(self.vault.decrypt(self._ciphertext))
 
     @data.setter
     def data(self, value):
-        self._ciphertext = value
+        self._ciphertext = to_bytes(value)
 
-    def __repr__(self):
-        return repr(self.data)
+    def is_encrypted(self):
+        return self.vault and self.vault.is_encrypted(self._ciphertext)
 
-    # Compare a regular str/text_type with the decrypted hypertext
     def __eq__(self, other):
         if self.vault:
             return other == self.data
         return False
 
-    def __hash__(self):
-        return id(self)
-
     def __ne__(self, other):
         if self.vault:
             return other != self.data
         return True
 
+    def __reversed__(self):
+        # This gets inerhited from ``collections.Sequence`` which returns a generator
+        # make this act more like the string implementation
+        return to_text(self[::-1], errors='surrogate_or_strict')
+
     def __str__(self):
         return to_native(self.data, errors='surrogate_or_strict')
 
     def __unicode__(self):
         return to_text(self.data, errors='surrogate_or_strict')
 
     def encode(self, encoding=None, errors=None):
-        return self.data.encode(encoding, errors)
+        return to_bytes(self.data, encoding=encoding, errors=errors)
+
+    # Methods below are a copy from ``collections.UserString``
+    # Some are copied as is, where others are modified to not
+    # auto wrap with ``self.__class__``
+    def __repr__(self):
+        return repr(self.data)
+
+    def __int__(self, base=10):
+        return int(self.data, base=base)
+
+    def __float__(self):
+        return float(self.data)
+
+    def __complex__(self):
+        return complex(self.data)
+
+    def __hash__(self):
+        return hash(self.data)
+
+    # This breaks vault, do not define it, we cannot satisfy this
+    # def __getnewargs__(self):
+    #    return (self.data[:],)
+
+    def __lt__(self, string):
+        if isinstance(string, AnsibleVaultEncryptedUnicode):
```

Comment:
Aren't these superfluous? The appropriate method will be called for each instance so you can just do
```suggestion
return self.data < string
```
Same for `__le__`, `__gt__`, etc.

Reply:
Everything after line 153 is a direct copy from `collections.UserString` and then modified to use `AnsibleVaultEncryptedUnicode` instead of `UserString` in isinstance calls, and to ensure they don't rewrap with `self.__class__`.